torch/distributed/_shard/sharded_tensor/_ops/_common.py

```diff
@@ -14,6 +21,171 @@
 )
 
 
+def _chunk_sharding_spec_check(spec, op):
+    """
+    For the given op implementation check if the sharding spec is ChunkShardingSpec.
+    """
+    if not isinstance(spec, shard_spec.ChunkShardingSpec):
+        raise NotImplementedError(
+            f"Only ChunkShardingSpec supported for '{op.__name__}'."
+        )
+
+
+def _sharded_op_common(op, early_stop_func, extra_check):
+    """
+    Inject sharded tensor op registration with common logics executed before
+    different behaviors are done on either local shards or a local tensor.
+
+    Example::
+        >>> op = torch.transpose
+        >>> @sharded_op_impl(op)
+        >>> @_sharded_op_common(op, early_stop_func, extra_check)
+        >>> def sharded_tensor_op(types, args, kwargs, process_group):
+        >>>   ....
+        >>>
+        >>> st = sharded_tensor.rand(32, 16)
+        >>> st.transpose(1, 2)
+        >>> # This will call '_sharded_op_common'
+
+    Args:
+        op: The op to be registered and applied to all shards of the st.
+        early_stop_func (Callable, optional): the func for early stop.
+            Default: if ``None``, no early stop.
+        extra_check (Callable, optional): the func for extra condition check.
+            Default: if ``None``, no extra check.
+
+    Return:
+        func (Callable): Torch function for which we want to provide a sharded
+            implementation (ex: torch.transpose)
+    """
+    def decorator_sharded_func(wrapped_func):
+        @functools.wraps(wrapped_func)
+        def wrapper(types, args=(), kwargs=None, pg=None):
+            if len(args) == 0:
+                raise ValueError(f" No input for '{op.__name__}'!")
+            # Validate types
+            st = args[0]
+            if not isinstance(st, ShardedTensor):
```

Comment:
I would image the first param is always a sharded tensor? Do you have a detailed use case example?

Reply:
hmm I think there's bunch of common use cases where args[0] is not sharded tensor that we need to consider, i.e. for a operation like `3 + st`(it will have args with the first one be `3` and the second one be `st), `replicated_tensor + st`, `, then args[0] won't be a ShardedTensor?

Reply:
Yeah i guess if this decorator is only used for the functions that registered in this PR, I think it make sense to separate them for now, we might need to consider the cases in the future.